### PR TITLE
Draft: encryption at rest

### PR DESCRIPTION
**Pull request by mberhault**
Some rudimentary description of encryption at rest, including full
example to:
* generate keys
* enable encryption
* change encryption

There is much left to do.

The section on the admin UI stores page requires
https://github.com/cockroachdb/cockroach/pull/26040

### CODE REVIEW

**Comment by mberhault** [comment]
CC: @nstewart, @rmloveland 

Here's the direct link to the generated page: http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/ec7ef6a82f633b8a2717bf825428b9adee3d8ba8/dev/encryption.html

Nate: this should be enough to get you started, although we still need https://github.com/cockroachdb/cockroach/pull/26040 to get merged (I'm hoping today).

Rich: I'll address the TODOs, there's quite a bit of background and explanation needed.

**Comment by mberhault** [comment]
The stores report PR is merged.

**Comment by mberhault** [comment]
FYI: this lives under: 2.1 -> Guides -> Deploy -> Enterprise Features -> Encryption At Rest.
Boy, we've really gone hierarchy crazy.

File: v2.1/encryption.md [hunk 1]
```diff
@@ -0,0 +1,127 @@
+---
+title: Encryption At Rest
```
Comment (rmloveland) [inline v2.1/encryption.md]: Same as previous comment - page title should end with "(Alpha)" or similar.

**Comment by nstewart** [comment]
Yes, at this point we just need enough docs to unblock pm acceptance testing and eng qa

**Comment by jseldess** [comment]
@nstewart, do you still want some initial docs for next Monday's alpha release as well? Or is that no longer a priority?

**Comment by rmloveland** [comment]
> Thanks for the comments Rich, but I just want to make sure we're on the same page: this is not even close to finished, and it's definitely not meant to be the final docs, more of a brain-dump on my part and as such will probably require some serious rewriting by people better at explaining than I am :)

I understand.  Sorry to comment too early, I wasn't sure how we usually work on these draft PRs.  I will stand back and wait for the go-ahead to review later or take over the PR when ready (or however it's done). :-)

**Comment by mberhault** [comment]
No worries. I'll let you know when I've done the majority of the page. Either today or tomorrow.

**Comment by mberhault** [comment]
The draft is pretty much ready. It will most likely need some serious improvements, but this should be a good start to understand how encryption at rest works, including examples and recommendations.

**Comment by rmloveland** [comment]
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[v2.1/encryption.md, line 20 at r4](https://reviewable.io/reviews/cockroachdb/docs/3193#-LDmim1iEiN89bs7lQVn:-LDmim1j9IYlb4vt3jqe:b-e3fx97) ([raw file](https://github.com/cockroachdb/docs/blob/08113eb6fb7172ffd465a942a7ef71b5ba76e5f0/v2.1/encryption.md#L20)):*
> ```Markdown
> 
> Encryption is performed in the [storage layer](architecture/storage-layer.html) and configured per store.
> All files used by the store, regarless of contents, are encrypted with the desired algorithm.
> ```

s/regarless/regardless/

---

*[v2.1/encryption.md, line 43 at r4](https://reviewable.io/reviews/cockroachdb/docs/3193#-LDmj9EBDHtZ_pEirUik:-LDmj9EBDHtZ_pEirUil:b-mzkp3p) ([raw file](https://github.com/cockroachdb/docs/blob/08113eb6fb7172ffd465a942a7ef71b5ba76e5f0/v2.1/encryption.md#L43)):*
> ```
> * prevent key reuse with the same encryption parameters (after many files)
> ```

Markdown isn't rendering properly for this list.  Maybe try an empty line before the list?

---

*[v2.1/encryption.md, line 50 at r4](https://reviewable.io/reviews/cockroachdb/docs/3193#-LDmjYUy2BpY-1Pl3CC0:-LDmjYUy2BpY-1Pl3CC1:b9de4yv) ([raw file](https://github.com/cockroachdb/docs/blob/08113eb6fb7172ffd465a942a7ef71b5ba76e5f0/v2.1/encryption.md#L50)):*
> ```Markdown
> 
> Data keys will automatically be rotated at startup if any of the following conditions are met:
> * the active store key changed
> ```

Same comment as above re: Markdown list rendering

---

*[v2.1/encryption.md, line 72 at r4](https://reviewable.io/reviews/cockroachdb/docs/3193#-LDmk4U498uWyFisq-sT:-LDmk4U498uWyFisq-sU:b-2f6spc) ([raw file](https://github.com/cockroachdb/docs/blob/08113eb6fb7172ffd465a942a7ef71b5ba76e5f0/v2.1/encryption.md#L72)):*
> ```Markdown
> ## Recommendations
> 
> There are a number of considerations to keep in mind when running with encryption:
> ```

s/:/./ since no list follows

---

*[v2.1/encryption.md, line 75 at r4](https://reviewable.io/reviews/cockroachdb/docs/3193#-LDmjgZ731Cc6Hwtj20d:-LDmjgZ731Cc6Hwtj20e:b-2c0wnn) ([raw file](https://github.com/cockroachdb/docs/blob/08113eb6fb7172ffd465a942a7ef71b5ba76e5f0/v2.1/encryption.md#L75)):*
> ```Markdown
> 
> Key management is the most dangerous aspect of encryption. The following rules should be kept in mind:
> * make sure only the unix user running the `cockroach` process has access to the keys
> ```

Markdown list render issue (same as above)

---

*[v2.1/encryption.md, line 81 at r4](https://reviewable.io/reviews/cockroachdb/docs/3193#-LDmjlit-CgJGw0beRF5:-LDmjlit-CgJGw0beRF6:b-2c0wnn) ([raw file](https://github.com/cockroachdb/docs/blob/08113eb6fb7172ffd465a942a7ef71b5ba76e5f0/v2.1/encryption.md#L81)):*
> ```Markdown
> 
> A few other recommendations apply for best security practices:
> * do not switch from encrypted to plaintext, this leaks data keys. Once transitioned to plaintext, all data must be considered reachable.
> ```

Markdown list render issue (same as above)

---

*[v2.1/encryption.md, line 168 at r4](https://reviewable.io/reviews/cockroachdb/docs/3193#-LDmkorO-dHTE59Nl9lR:-LDmkorO-dHTE59Nl9lS:b-43qtdl) ([raw file](https://github.com/cockroachdb/docs/blob/08113eb6fb7172ffd465a942a7ef71b5ba76e5f0/v2.1/encryption.md#L168)):*
> ```
> key=/path/to/my/old-aes-128.key
> ```
do you mean `old-key=/path/to/my/old-aes-128.key`?

---

*[v2.1/encryption.md, line 173 at r4](https://reviewable.io/reviews/cockroachdb/docs/3193#-LDml3i_0zu459f_NApo:-LDml3i_0zu459f_NApp:bc5f3vz) ([raw file](https://github.com/cockroachdb/docs/blob/08113eb6fb7172ffd465a942a7ef71b5ba76e5f0/v2.1/encryption.md#L173)):*
> ```Markdown
> ## See Also
> 
> TODO(mberhault): links to external resources, report page details, flag descriptions.
> ```

Could replace this with something simple for now e.g.

```
- [Enterprise Licensing](enterprise-licensing.html)
```

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/docs/3193)*

**Comment by rmloveland** [comment]
@mberhault if you prefer, I can take over the PR and finish making edits

**Comment by mberhault** [comment]
Yes please, I'm deep into debugging an issue with encryption, I won't have time to look at this again today or tomorrow.

**Comment by mberhault** [comment]
Review status: 0 of 2 files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---

*[v2.1/encryption.md, line 84 at r4](https://reviewable.io/reviews/cockroachdb/docs/3193#-LDqLIQn5U1edkJTmx0B:-LDqNSYsCj0pw_Um8dha:b-p4a87g) ([raw file](https://github.com/cockroachdb/docs/blob/08113eb6fb7172ffd465a942a7ef71b5ba76e5f0/v2.1/encryption.md#L84)):*
<details><summary><i>Previously, nstewart (Nate) wrote…</i></summary><blockquote>

@mberhault -- we need to include a sentence or two for dealing with backups. What happens when I take a backup of a secure cluster? If I want to encrypt backups, what would be our recommendation? Just use encrypted s3? zip up the files and then encrypt that? 

cc @rmloveland
</blockquote></details>

Backup (as in our distributed backup) still works as expected and does not encrypt anything. Encrypting the actual backups will be a completely different feature that likely reuses nothing from this. Ditto for import/export/dump.

Copying raw files from disk when encryption is turned on is now impossible (it's mentioned in the doc).

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/docs/3193)*

**Comment by nstewart** [comment]
*[v2.1/encryption.md, line 84 at r4](https://reviewable.io/reviews/cockroachdb/docs/3193#-LDqLIQn5U1edkJTmx0B:-LDqP7Ya58C88BDE5MQ4:b-nculrl) ([raw file](https://github.com/cockroachdb/docs/blob/08113eb6fb7172ffd465a942a7ef71b5ba76e5f0/v2.1/encryption.md#L84)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

Backup (as in our distributed backup) still works as expected and does not encrypt anything. Encrypting the actual backups will be a completely different feature that likely reuses nothing from this. Ditto for import/export/dump.

Copying raw files from disk when encryption is turned on is now impossible (it's mentioned in the doc).
</blockquote></details>

Right -- I'm not talking about copying raw files, I'm talking about working with the files created by the distributed (unencrypted) backup. My question is, even if we don't have a feature for encrypting backups, what would be our recommendation for people that want to encrypt the backed up files? 

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/docs/3193)*

**Comment by mberhault** [comment]
Review status: 0 of 2 files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---

*[v2.1/encryption.md, line 84 at r4](https://reviewable.io/reviews/cockroachdb/docs/3193#-LDqLIQn5U1edkJTmx0B:-LDqQ8jjBImhNBvrgWVp:b-5m7rvc) ([raw file](https://github.com/cockroachdb/docs/blob/08113eb6fb7172ffd465a942a7ef71b5ba76e5f0/v2.1/encryption.md#L84)):*
<details><summary><i>Previously, nstewart (Nate) wrote…</i></summary><blockquote>

Right -- I'm not talking about copying raw files, I'm talking about working with the files created by the distributed (unencrypted) backup. My question is, even if we don't have a feature for encrypting backups, what would be our recommendation for people that want to encrypt the backed up files? 
</blockquote></details>

It depends entirely on whether they trust the storage used as the backup destination. If they don't trust AWS/GCE/etc..., they'll need to encrypt the backups themselves. I don't think we want to provide more guidance than that though.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/docs/3193)*

**Comment by nstewart** [comment]
*[v2.1/encryption.md, line 84 at r4](https://reviewable.io/reviews/cockroachdb/docs/3193#-LDqLIQn5U1edkJTmx0B:-LDqQonF1zSn31VmLSzs:b-2234oh) ([raw file](https://github.com/cockroachdb/docs/blob/08113eb6fb7172ffd465a942a7ef71b5ba76e5f0/v2.1/encryption.md#L84)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

It depends entirely on whether they trust the storage used as the backup destination. If they don't trust AWS/GCE/etc..., they'll need to encrypt the backups themselves. I don't think we want to provide more guidance than that though.
</blockquote></details>

I think that's enough. Basically just saying: hey, backups/exports/etc aren't encrypted; you will have to encrypt those using your favorite encryption method.

cc @rmloveland

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/docs/3193)*

**Comment by rmloveland** [comment]
Review status: 0 of 2 files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---

*[v2.1/encryption.md, line 84 at r4](https://reviewable.io/reviews/cockroachdb/docs/3193#-LDqLIQn5U1edkJTmx0B:-LDqzqQB0TaNqdpPWG8m:b-s5069d) ([raw file](https://github.com/cockroachdb/docs/blob/08113eb6fb7172ffd465a942a7ef71b5ba76e5f0/v2.1/encryption.md#L84)):*
> I think that's enough. Basically just saying: hey, backups/exports/etc aren't encrypted; you will have to encrypt those using your favorite encryption method.

Just added a note to that effect.  Thanks Nate!

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/docs/3193)*

**Comment by rmloveland** [comment]
:lgtm:

---

Review status: 0 of 2 files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/docs/3193#-:-LDr0tHEAJs2IhD9NN7j:bnfp4nl)*

File: v2.1/encryption.md [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+followed by the key (16, 24, or 32 bytes). The size of the key dictates the version of AES to use (AES-128, AES-192, or AES-256).
+
+At startup, CockroachDB uses a data key with the same length as the store key. If encryption has just been enabled,
+the key size has changed, or the data key is too old (default lifetime is one week), CockroachDB generates a new data key.
+
+Any new file created by the store uses the currently-active data key. All data keys (both active and previous) are stored in a key registry file and encrypted with the active store key.
+
+CockroachDB does not currently force re-encryption of older files but instead relies on normal RocksDB churn to slowly rewrite all data with the desired encryption.
+
+## Key rotation
```
[the remark this answers is not in the document]
Comment (rmloveland) [reply v2.1/encryption.md]: Fixed

File: v2.1/encryption.md [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+---
+
+<span class="version-tag">New in v2.1:</span>
+Encryption At Rest provides transparent encryption for node data on local disk.
+
+{% include experimental-warning.md %}
+
+<div id="toc"></div>
+
+## Background
```
[the remark this answers is not in the document]
Comment (rmloveland) [reply v2.1/encryption.md]: Fixed!

File: v2.1/encryption.md [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+The `key` and `old-key` components must **always** be specified. They allow for transitions between
+encryption algorithms, and between plaintext and encrypted.
+
+Starting a node for the first time using AES-128 encryption can be done using:
+{% include copy-clipboard.html %}
+~~~ shell
+$ cockroach start --store=cockroach-data --enterprise-encryption=path=cockroach-data,key=/path/to/my/aes-128.key,old-key=plain
+~~~
+
+**WARNING**: once specified for a given store, the `--enterprise-encryption` flag must always be present.
```
[the remark this answers is not in the document]
Comment (rmloveland) [reply v2.1/encryption.md]: Fixed!

File: v2.1/encryption.md [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+The key file must contain random data making up the key ID (32 bytes) and the actual key (16, 24, or 32
+bytes depending on the encryption algorithm).
+
+| Algorithm | Key size | Key file size |
+|-|-|-|
+| AES-128 | 128 bits (16 bytes) | 48 bytes |
+| AES-192 | 192 bits (24 bytes) | 56 bytes |
+| AES-256 | 256 bits (32 bytes) | 64 bytes |
+
+Generating a key file can be done using the `cockroach` CLI::
```
[the remark this answers is not in the document]
Comment (rmloveland) [reply v2.1/encryption.md]: Fixed!

**Pull request merged**